DBZ-2543 Document Oracle OpenLogReplicator adapter (#4775)
debezium/debezium: documentation/modules/ROOT/pages/connectors/oracle.adoc

|===

[[oracle-openlogreplicator-support]]
== OpenLogReplicator support

The {prodname} Oracle connector by default ingests changes using native Oracle LogMiner.
However, the connector can be toggled to use OpenLogReplicator, an open-source and free third-party application that reads Oracle changes directly from the redo and archive logs with low impact on the database.

> To configure the connector to use OpenLogReplicator, you must apply specific database and connector configurations that differ from those that you use with LogMiner.

Reviewer: ```suggestion
To configure the connector to use OpenLogReplicator, you must apply specific database and connector configurations that differ from the default LogMiner settings.
```

.Prerequisites

> * Download and compile OpenLogReplicator for your database environment.
> * OpenLogReplicator must be installed with direct access to the archive and redo log files. This does not necessarily require installation on the physical database server if the archive and redo logs can be accessed via some shared filesystem.

Reviewer: ```suggestion
* You downloaded and compiled OpenLogReplicator for your database environment. 
* OpenLogReplicator is installed with direct access to the archive and redo log files. 
Installation on the physical database server is not required, as long as the application can access archive and redo logs through a shared file system.
```

=== How OpenLogReplicator works

> OpenLogReplicator takes on the role of Oracle LogMiner and Oracle XStream when the {prodname} Oracle connector streams changes.

Reviewer: ```suggestion
OpenLogReplicator takes on the role of Oracle LogMiner and Oracle XStream as the {prodname} Oracle connector streams changes.
```

> It is responsible for the capturing of changes in the redo and archive logs as they occur and batching those changes into logical transactions.

Reviewer: ```suggestion
The application captures changes from the redo and archive logs as they occur, and batches those changes into logical transactions.
```

> {prodname} Oracle connector is a consumer of OpenLogReplicator by connecting to the network endpoint provided by OpenLogReplicator and ingesting the transactions as they're batched.

Reviewer: ```suggestion
OpenLogReplicator provides a network endpoint for consuming applications. 
After you connect the {prodname} Oracle connector to the provided endpoint, it ingests transactions from OpenLogReplicator as they are batched.
```

Once a change has been ingested by the OpenLogReplicator adapter, this {prodname} connector transforms the events into xref:#oracle-events[Data change events] just like any other adapter.

From a network topology perspective, the {prodname} Oracle connector relies on a network connection to both Oracle and OpenLogReplicator.
Similarly, OpenLogReplicator requires a network connection to the Oracle database along with direct access to the raw redo and archive logs.

=== Prepare the database

The Oracle database must be configured to generate specific files called archive logs.
An archive log is a saved copy of a full online redo log.
Archive logs are only generated when the database is placed into `ARCHIVELOG` mode and this mode requires that specific configuration properties by set to identify where these archive log files will be stored.

The following illustrates the steps to enable archive log mode.

.Configuration needed for OpenLogReplicator

> [source,indent=0]

Author: I'm not sure I understand this specific change, can you elaborate @roldanbob ?

> [source,indent=0]

Reviewer: The intention is to set off the example in an example block, which automatically inserts an example number in the rendered doc.

> [source,indent=0]

Author: That makes sense, I wasn't aware of this formatting tbh, so it caught me off guard.  Thanks for explaining.

----
ORACLE_SID=ORCLCDB dbz_oracle sqlplus /nolog

CONNECT sys/top_secret AS SYSDBA
alter system set db_recovery_file_dest_size = 5G;
alter system set db_recovery_file_dest = '/opt/oracle/oradata/recovery_area' scope=spfile;
shutdown immediate
startup mount
alter database archivelog;
alter database open;
-- Should show "Database log mode: Archive Mode"
archive log list

exit;
----

> 

Author: Similar to the above.

> 

Reviewer: The `====` closes the example block.

In addition, supplemental logging is required for Oracle to write the necessarily pieces of information about what has changed to the redo logs so that {prodname} can generate change events with before and after state.
Supplemental logging can be configured a number of different ways, but {prodname} for Oracle only requires minimal logging to work.

The database must be configured with minimal supplemental logging.
Minimal supplemental logging writes the least amount of information needed to be able to create change events.
The following illustrates how to enable minimal supplemental logging.

.Configure database supplemental logging
[source,sql,indent=0]
----
ALTER DATABASE ADD SUPPLEMENTAL LOG DATA
----

> 

Author: Similar to the above.

> 

Reviewer: Same response as for L4244.

[NOTE]
====
{prodname} for Oracle can work with a higher degree of supplemental logging if that is already configured for other purposes.
If you are replacing a prior solution that required a higher fidelity of logging with {prodname}, you can chose to reduce the logging levels accordingly as minimum is all that is required at the database level.
====

> 

Author: While I don't disagree, I wonder if we shouldn't do that as a follow-up and align LogMiner, XStream, and the new OpenLogReplicator sections to match as the current layout was primarily taken from what we had as a layout for the other two.  wdyt?

> 

Reviewer: Yes, I agree. Certainly not urgent. It was just something that occurred to me as I went through and tried to understand the content.

> 

Author: I'd say add a Jira so we don't loose sight of it and we can come back to it as time permits.

> 

Reviewer: DBZ-6818

Finally, each captured table must be explicitly configured with a higher fidelity of supplemental logging called `(ALL) COLUMNS`.
This logging level guarantees that Oracle captures the state of a column regardless of whether the column changed when a redo entry is written to the redo log.
This allows {prodname} for Oracle to generate change events safely with both _before_ and _after_ state.

The following illustrates how to enable supplemental logging for each captured table.

.Configure captured table supplemental logging

> [source,sql,indent=0]

Author: Right, but shouldn't this actually be:
```
=== Configuring supplemental logging for a captured table
[source,sql,indent=0]
```
I also wonder if we shouldn't think about adding some text before the source block.  I'm not a huge fan of adding a section heading which is then immediately followed by a source box; wdyt?

> [source,sql,indent=0]

Reviewer: We could do a section heading, but I think that implementing the example block, which converts the existing title text into an Example heading might be sufficient. See *Example 3.1* in the [Db2 Deployment section of the user guide](https://access.redhat.com/documentation/en-us/red_hat_integration/2023.q2/html-single/debezium_user_guide/index#openshift-streams-db2-connector-deployment). T 
In general, I agree that adding some intervening text to describe what's in the example could be helpful. I'm not sure whether it's necessary in this instance?

        "database.port" : "1521",
        "database.user" : "c##dbzuser",
        "database.password" : "dbz",
        "database.dbname" : "ORCLCDB",
        "database.pdb.name" : "ORCLPDB1",
        "schema.history.internal.kafka.bootstrap.servers" : "kafka:9092",
        "schema.history.internal.kafka.topic": "schema-changes.inventory",
        "database.connection.adapter": "olr",
        "openlogreplicator.source": "ORACLE",
        "openlogreplicator.host": "<ip address or hostname of OpenLogReplicator>",
        "openlogreplicator.port": "<port OpenLogReplicator is listening on>"
    }
}
----

[id="building-openlogreplicator"]
=== Building OpenLogReplicator

OpenLogReplicator does not distribute binaries, so you must build the third-party tool from code.
It is written in C++; however, the author provides a container image to build the source for a variety of operating systems such as Debian, Centos, Fedora, etc.
This https://github.com/bersler/OpenLogReplicator-docker[repository] provides details on how to build the tool using a container.

[id="obtaining-oracle-jdbc-driver-openlogreplicator"]
=== Obtaining the Oracle JDBC driver for OpenLogReplicator

> The {prodname} Oracle connector requires the Oracle JDBC driver to connect to Oracle databases.
> Please refer to the link:#obtaining-the-oracle-jdbc-driver[Obtaining the Oracle JDBC driver] as OpenLogReplicator has the exact same requirements as Oracle LogMiner.

Reviewer: ```suggestion
When you use the {prodname} Oracle connector with the OpenLogReplicator, you must  obtain the Oracle JDBC driver to connect to Oracle databases.
For more information, see xref:obtaining-the-oracle-jdbc-driver[Obtaining the Oracle JDBC driver].
```

[[oracle-openlogreplicator-configuration]]
=== OpenLogReplicator configuration

OpenLogReplicator is a third-party standalone process that is responsible for reading the Oracle redo and archive logs.

> Unlike other {prodname} Oracle adapters, this means that there are two configurable components that must be configured independently of one another.

Reviewer: ```suggestion
Unlike other {prodname} Oracle adapters, OpenReplicator includes two separate components, which you must configure independently of each other.
```

Reviewer: I'm confused by the reference to the _two configurable components_. Along with the configuration described in this section, I assume that the sentence refers to the configuration described in `selecting-the-adapter`, which, I think is really less about configuring the component and more about specifying that you want to use the OpenLogReplicator.

If that's correct, we should move this sentence to earlier in the doc (e.g., L4357?) and refactor the topic _Configuring the OpenLogReplicator adapter_ so that it includes the subtopics _Selecting the adapter_, and _Configuring the adapter_. 
If that's not correct, I'm not sure what the second component is.

Or maybe just remove that sentence?

Author: What I am referring to is that unlike other connectors where you focus on providing a Debezium connector config that is used by Kafka Connect, Debezium Server, or the Embedded Engine to run the connector, the user must also correctly configure and start the OpenLogReplicator process.  In other words, this additional configuration file must be created and set with a rigid set of values for the integration to work.  

I'm open to how you'd like best to address this clearly and concisely.

Reviewer: Ok, I think that I get it.  Maybe I was just getting tripped up by the adapter terminology, which is new to me.  I also don't know much about Debezium server or the Embedded Engine. 
In any case, I think it might better if we just avoid making the negative comparison (_"Unlike ..."_), and simply say, "To use the {prodname} Oracle connector with the OpenReplicator adapter, you must provide a configuration for the adapter that is separate from the xref:oracle-adding-connector-configuration[configuration that you specify for the connector itself]."

We don't really have a separate Oracle connector configuration topic for downstream, adding the link won't work in that context, but we can either punt on that for now, or leave out the link.

Author: > I'll apply that edit to L4418 so you can either commit it, or push it back if you think that it still needs work.

That's fine, feel free to pull down this branch and push a commit to it if you want; entirely up to you.

> We don't really have a separate Oracle connector configuration topic for downstream, so adding the xref won't work in that context, so, for now, we can either punt on doing any conditionalization of the sentence, or leave out the link.

I think since OpenLogReplicator isn't going downstream anytime soon, we can leave it as is for now with the link and we can revisit this when we "downstreamify" this huge piece of work.

> OpenLogReplicator is configured by using a JSON-based configuration in a file called `scripts/OpenLogReplicator.json`.

Reviewer: ```suggestion
You configure OpenLogReplicator by using a JSON file called `scripts/OpenLogReplicator.json`.
```

> The format of this JSON configuration can be found https://github.com/bersler/OpenLogReplicator/blob/master/documentation/reference-manual/reference-manual.adoc#openlogreplicator-json-file-format[here].

Reviewer: ```suggestion
For more information about the required format of this file, see the https://github.com/bersler/OpenLogReplicator/blob/master/documentation/reference-manual/reference-manual.adoc#openlogreplicator-json-file-format[OpenLogReplicator documentation].
```

> .OpenLogReplicator configuration
> [source,json,indent=0]

Reviewer: ```suggestion
.OpenLogReplicator configuration
====
[source,json,indent=0]
```

----
{
  "version": "1.2.1",
  "source": [{
    "alias": "SOURCE",

>     "name": "ORACLE", <.>
>     "reader": {
>       "type": "online",
>       "path-mapping": ["/opt/olr/recovery_area", "/opt/olr/recovery_area/ORCLCDB/archivelog"], <.>
>       "user": "c##dbzuser", <.>
>       "password": "dbz", <.>
>       "server": "//<ip>:<port>/ORCLPDB1" <.>
>     },
>     "format": { <.>
>       "type": "json",
>       "column": 2,
>       "db": 3,
>       "interval-dts": 9,
>       "interval-ytm": 4,
>       "message": 2,
>       "rid": 1,
>       "schema": 7,
>       "scn-all": 1,
>       "timestamp-all": 1
>     }
>   }],
>   "target": [{
>     "alias": "DEBEZIUM",
>     "source": "SOURCE",
>     "writer": {
>       "type": "network", <.>
>       "uri": "<host>:<port>" <.>

Reviewer: ```suggestion
    "name": "ORACLE", <1>
    "reader": {
      "type": "online",
      "path-mapping": ["/opt/olr/recovery_area", "/opt/olr/recovery_area/ORCLCDB/archivelog"], <2>
      "user": "c##dbzuser", <3>
      "password": "dbz", <4>
      "server": "//<ip>:<port>/ORCLPDB1" <5>
    },
    "format": { <6>
      "type": "json",
      "column": 2,
      "db": 3,
      "interval-dts": 9,
      "interval-ytm": 4,
      "message": 2,
      "rid": 1,
      "schema": 7,
      "scn-all": 1,
      "timestamp-all": 1
    }
  }],
  "target": [{
    "alias": "DEBEZIUM",
    "source": "SOURCE",
    "writer": {
      "type": "network", <7>
      "uri": "<host>:<port>" <8>
```

    }
  }]
}
----

> <.> This should match the `openlogreplicator.source` connector configuration.
> <.> List of file path pairs `[before1,after1,befor2,after2,...]` where if a log file path matches one of the `beforeX` prefixes, the prefix is replaced with the `afterX` path.  This is useful when OpenLogReplicator is running on a different host than the source database and the path to the redo and archive logs different between the database and the OpenLogReplicator process.
> <.> This should match the `database.user` connector property.
> <.> This should match the `database.password` connector property.
> <.> This should point to the database host, port, and Oracle SID.
> <.> This specifies the payload format ingested by {prodname} Oracle connector. Use these values as specified, as these are the only format options that we require beyond the defaults.
> <.> This must specify `network`, as {prodname} Oracle connector communicates with OpenLogReplicator via a network connection.
> <.> This specifies the bind host and port that OpenLogReplicator listens for connections on.  This should be accessible by {prodname} Oracle connector.

Reviewer: ```suggestion
<1> This should match the `openlogreplicator.source` connector configuration.
<2> List of file path pairs `[before1,after1,befor2,after2,...]` where if a log file path matches one of the `beforeX` prefixes, the prefix is replaced with the `afterX` path.  This is useful when OpenLogReplicator is running on a different host than the source database and the path to the redo and archive logs different between the database and the OpenLogReplicator process.
<3> This should match the `database.user` connector property.
<4> This should match the `database.password` connector property.
<5> This should point to the database host, port, and Oracle SID.
<6> This specifies the payload format ingested by {prodname} Oracle connector. Use these values as specified, as these are the only format options that we require beyond the defaults.
<7> This must specify `network`, as {prodname} Oracle connector communicates with OpenLogReplicator via a network connection.
<8> This specifies the bind host and port that OpenLogReplicator listens for connections on.  This should be accessible by {prodname} Oracle connector.
```

Although the period characters work in the rendered doc, it's best to explicitly identify each callout with a numeral, so that you can easily collate them in the asciidoc source.

> <.> This specifies the bind host and port that OpenLogReplicator listens for connections on.  This should be accessible by {prodname} Oracle connector.
> 

Reviewer: ```suggestion
<.> This specifies the bind host and port that OpenLogReplicator listens for connections on.  This should be accessible by {prodname} Oracle connector.
====

```

Add closing example block delimiter and a blank line to precede the next heading.

[[oracle-openlogreplicator-connector-properties]]
=== OpenLogReplicator connector properties

> The following configuration properties are _required_ when using OpenLogReplicator unless a default value is available.

Reviewer: ```suggestion
The following configuration properties are required when using OpenLogReplicator, unless a default value is available.
```

The table indicates that none of these properties have default values. Can we remove the last part of the preceding sentence? Or should we keep it in case we later implement defaults?

Author: I think we can remove it for now; however, I want to point out that we use the `_required_` italic text elsewhere in the documentation.  Should we also adjust those other lines?

Reviewer: Technically, I don't think that it's necessary, but as a rule, I like to err on the side of minimizing the use of text formatting to emphasize content.

Author: To be clear, what I meant was in all our other connector properties sections in the Oracle documentation, we use the same phrasing with the same formatting.  Should we align those to be uniform?

Reviewer: Ok, got it. No, let's just leave everything as-is for now. It really ain't broke, so let's not fix it.

[cols="30%a,25%a,45%a"]
|===
|Property
|Default
|Description

|[[oracle-property-openlogreplicator-source]]<<oracle-property-openlogreplicator-source, `+openlogreplicator.source+`>>
|No default
|The logical name of the configured `source.name` element in the OpenLogReplicator JSON configuration.

|[[oracle-property-openlogreplicator-host]]<<oracle-property-openlogreplicator-host, `+openlogreplicator.host+`>>
|No default

> |The host or IP of the OpenLogReplicator network service.

Reviewer: ```suggestion
|The host name or IP address of the OpenLogReplicator network service.
```

|[[oracle-property-openlogreplicator-port]]<<oracle-property-openlogreplicator-port, `+openlogreplicator.port+`>>
|No default

> |The port of the OpenLogReplicator network service.

Reviewer: ```suggestion
|The port number that is used by the OpenLogReplicator network service.
```
